### DISCUSSION TITLE
Using a Razor Class Library inside an Orchard Core Module

### DISCUSSION THREAD

**Discussion by max-roeder** [OrchardCMS/OrchardCore#17056] [Q&A] [answered]
Hello everyone,

i hope you can help me. As the Title suggests i have a Module in which i plan to use a RCL which is also used in another ASP.NET Core application. When creating a ViewComponent inside the Module everything works fine. When i reference the RCL and try to use a ViewComponent i get an InvalidOperationException which states, that the ViewComponent could not be found. The Other Application i mentioned (Razor Pages) does not have any problems. I put together some Sample Code to show you: https://github.com/max-roeder/OrchardRCLTest.

**Comment by hishamco** [comment]
Why not using your views directly from within the OC Module itself?

**Reply by max-roeder** [reply]
By Views do you mean the ViewComponents? Like i said i want to use the ViewComponents in another application. 
Or do you mean, since Orchard Modules themselves are RCLs i should use them as RCL and use the ViewComponents from the OrchardModule in my other Application? Since the Other Application is a "Standalone" Razor Pages project i would like to keep Orchard Dependencies out of my RCL.

**Reply by hishamco** [reply]
Why must you separate the views or view components into another project? I presume the view components are related to the host app, or do you use them for something else?

**Reply by max-roeder** [reply]
I have the following Structure:
- Orchard Module, should reference RCL
- Razor Pages Project which also should reference RCL
- RCL which contains ViewComponents i want to use in both Orchard and Razor Pages project

The Razor Pages project should live beside the Orchard Application, these are two different Applications, but they "live" in the same space. They belong to the Same Product. Basically the Orchard Module is mainly focused on Marketing purposes, so Marketing has a tool to create a Website around Product Informations, so we dont have to implement CMS functionality ourselves in our Razor Pages project which is focused towards our main user group. But we have some View Components in our Razor Pages which we also need in the Orchard Module so I put them inside an RCL so we dont need to maintain 2 different sets of ViewComponents which do the same. I hope this clarifies things.

**Comment by max-roeder** [comment, marked as answer]
I finally managed to get this working. After Adding Mvc and the ApplicationPart for my Assembly explicitly and adding an EmbeddedFileProvider for my AddRazorRuntimeCompilation i only needed to tweak the path in my ViewComponent it now finds the ViewComponent and the corresponding View. I updated everything on my Sample Application  https://github.com/max-roeder/OrchardRCLTest.

**Reply by aaronamm** [reply]
@max-roeder  thanks for sharing your sample project.
